sw/airborne/modules/dcf/dcf.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) 2017 Hector Garcia de Marina
+ *
+ * This file is part of paparazzi.
+ *
+ * paparazzi is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ */
+
+#include <math.h>
+#include <std.h>
+#include <stdio.h>
+
+#include "modules/dcf/dcf.h"
+#include "subsystems/datalink/datalink.h" // dl_buffer
+#include "subsystems/datalink/telemetry.h"
+#include "subsystems/navigation/common_nav.h"
+#include "autopilot.h"
+
+#if PERIODIC_TELEMETRY
+static void send_dcf(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_DCF(trans, dev, AC_ID, 4 * DCF_MAX_NEIGHBORS, &(tableNei[0][0]), DCF_MAX_NEIGHBORS, error_sigma);
+}
+#endif // PERIODIC TELEMETRY
+
+// Control
+/*! Default gain k for the algorithm */
+#ifndef DCF_GAIN_K
+#define DCF_GAIN_K 10
+#endif
+/*! Default radius for the circumference */
+#ifndef DCF_RADIUS
+#define DCF_RADIUS 80
+#endif
+/*! Default timeout for the neighbors' information */
+#ifndef DCF_TIMEOUT
+#define DCF_TIMEOUT 1500
+#endif
+
+dcf_con dcf_control = {DCF_GAIN_K, DCF_RADIUS, DCF_TIMEOUT, 0};
+
+/*! Default number of neighbors per aircraft */
+#ifndef DCF_MAX_NEIGHBORS
+#define DCF_MAX_NEIGHBORS 4
+#endif
+int16_t tableNei[DCF_MAX_NEIGHBORS][4];
```

Comment:
Can you make this an array of structs with elements containing nei_id, theta, desired_sigma and the last_theta (merge the last_theta and tableNei). It will make the use of the elements much more intuitive.

Additionally, consider using more descriptive variable names? Theta and sigma are not very intuitive without any explanation. Another one last_theta -> msg_time

It also doesn't look like the tableNei[i][3] is used anywhere, why do you store the timeout?

Reply:
I agree that it would be more intuitive. However, I believe it will make the code more cumbersome in many places. What about a comment on top of this line describing each element? (they are also described in the wiki).

Sigma and theta are described in the wiki and in the corresponding papers. I will write a comment in the code and put the reference to the wiki.

I will do the same with last_theta (last time that the aircraft received a theta).

The timeout is sent in a telemetry message. An operator in the ground can monitor the frequency of lost messages and decide to have a longer timeout (if a neighbor is in 'timeout' then it left the formation).